Feature: Add support for function-based completion in 'complete' option

girishji opened the pull request:
This change adds two new values to the `'complete'` (`'cpt'`) option:
- `f` – invokes the function specified by the `'completefunc'` option
- `f{func}` – invokes a specific function `{func}` (can be a string or `Funcref`)

These new flags extend keyword completion behavior (e.g., via `<C-N>` / `<C-P>`) by allowing function-based sources to participate in standard keyword completion.

**Key behaviors:**

- Multiple `f{func}` values can be specified, and all will be called in order.
- Functions should follow the interface defined in `:help complete-functions`.
- When using `f{func}`, escaping is required for spaces (with `\`) and commas (with `\\`) in `Funcref` names.
- If a function sets `'refresh'` to `'always'`, it will be re-invoked on every change to the input text. Otherwise, Vim will attempt to reuse and filter existing matches as the input changes, which matches the default behavior of other completion sources.
- Matches are inserted at the keyword boundary for consistency with other completion methods.
- If finding matches is time-consuming, `complete_check()` can be used to maintain responsiveness.
- Completion matches are gathered in the sequence defined by the `'cpt'` option, preserving source priority.

This feature increases flexibility of standard completion mechanism and may reduce the need for external completion plugins for many users.

**Examples:**

Complete matches from [LSP](https://github.com/yegappan/lsp) client. Notice the use of `refresh: always` and `function()`.

```vim
set cpt+=ffunction("g:LspCompletor"\\,\ [5]). # maxitems = 5

def! g:LspCompletor(maxitems: number, findstart: number, base: string): any
    if findstart == 1
        return g:LspOmniFunc(findstart, base)
    endif
    return {words: g:LspOmniFunc(findstart, base)->slice(0, maxitems), refresh: 'always'}
enddef
autocmd VimEnter * g:LspOptionsSet({ autoComplete: false, omniComplete: true })
```

Complete matches from `:iabbrev`.

```vim
set cpt+=fAbbrevCompletor

def! g:AbbrevCompletor(findstart: number, base: string): any
    if findstart > 0
        var prefix = getline('.')->strpart(0, col('.') - 1)->matchstr('\S\+$')
        if prefix->empty()
            return -2
        endif
        return col('.') - prefix->len() - 1
    endif
    var lines = execute('ia', 'silent!')
    if lines =~? gettext('No abbreviation found')
        return v:none  # Suppresses warning message
    endif
    var items = []
    for line in lines->split("\n")
        var m = line->matchlist('\v^i\s+\zs(\S+)\s+(.*)$')
        if m->len() > 2 && m[1]->stridx(base) == 0
            items->add({ word: m[1], info: m[2], dup: 1 })
        endif
    endfor
    return items->empty() ? v:none :
        items->sort((v1, v2) => v1.word < v2.word ? -1 : v1.word ==# v2.word ? 0 : 1)
enddef
```

**Auto-completion:**

Vim's standard completion frequently checks for user input while searching for new matches. It is responsive irrespective of file size. This makes it well-suited for smooth auto-completion. You can try with above examples:

```vim
set cot=menuone,popup,noselect inf

autocmd TextChangedI * InsComplete()

def InsComplete()
    if getcharstr(1) == '' && getline('.')->strpart(0, col('.') - 1) =~ '\k$'
        SkipTextChangedIEvent()
        feedkeys("\<c-n>", "n")
    endif
enddef

inoremap <silent> <c-e> <c-r>=<SID>SkipTextChangedIEvent()<cr><c-e>

def SkipTextChangedIEvent(): string
    set eventignore+=TextChangedI  # Suppress next event caused by <c-e> (or <c-n> when no matches found)
    timer_start(1, (_) => {
        set eventignore-=TextChangedI
    })
    return ''
enddef

# Optional
inoremap <silent><expr> <tab> pumvisible() ? "\<c-n>" : "\<tab>"
inoremap <silent><expr> <s-tab> pumvisible() ? "\<c-p>" : "\<s-tab>"
```

girishji committed:
Add support for function-based completion in 'complete' option

This change adds two new values to the `'complete'` (`'cpt'`) option:
- `f` – invokes the function specified by the `'completefunc'` option
- `f{func}` – invokes a specific function `{func}` (can be a string or
  `Funcref`)

These new flags extend keyword completion behavior (e.g., via `<C-N>` /
`<C-P>`) by allowing function-based sources to participate in standard keyword
completion.

Key behaviors:
- Multiple `f{func}` values can be specified, and all will be called in order.
- Functions should follow the interface defined in `:help complete-functions`.
- When using `f{func}`, escaping is required for spaces (with `\`) and commas
  (with `\\`) in `Funcref` names.
- If a function sets `'refresh'` to `'always'`, it will be re-invoked on every
  change to the input text. Otherwise, Vim will attempt to reuse and filter
  existing matches as the input changes, which matches the default behavior of
  other completion sources.
- Matches are inserted at the keyword boundary for consistency with other
  completion methods.
- If finding matches is time-consuming, `complete_check()` can be used to
  maintain responsiveness.
- Completion matches are gathered in the sequence defined by the `'cpt'`
  option, preserving source priority.

This feature increases flexibility of standard completion mechanism and may
reduce the need for external completion plugins for many users.

M  runtime/doc/insert.txt
M  runtime/doc/options.txt
M  src/insexpand.c
M  src/optionstr.c
M  src/testdir/test_ins_complete.vim
M  src/testdir/test_options.vim

girishji committed:
Fix compiler warnings

M  src/insexpand.c

girishji committed:
Fix Windows compile errors

M  src/insexpand.c

girishji committed:
Fix warnings

M  src/insexpand.c

girishji committed:
Remove warnings

M  src/insexpand.c

girishji committed:
Fix warnings

M  runtime/doc/options.txt
M  src/insexpand.c

girishji committed:
Remove warning

M  src/insexpand.c

girishji committed:
Fix warnings

M  src/insexpand.c

girishji committed:
Merge branch 'vim:master' into pr_func_compl

girishji committed:
Merge branch 'vim:master' into pr_func_compl

girishji committed:
Improve doc

M  runtime/doc/options.txt

girishji committed:
Add support for 'o' flag in 'complete' option to invoke 'omnifunc' function

M  runtime/doc/options.txt
M  src/insexpand.c
M  src/optionstr.c
M  src/testdir/test_ins_complete.vim
M  src/testdir/test_options.vim

chrisbra commented on runtime/doc/options.txt:
```diff
@@ -2085,6 +2085,28 @@ A jump table for the options with a short description can be found at |Q_op|.
+		is invoked and what it should return.  The value can be the
+		name of a function or a |Funcref|.  For |Funcref| values,
+		spaces must be escaped with a backslash ('\'), and commas with
+		double backslashes ('\\').
```
```suggestion
		double backslashes ('\\') (see |option-backslash|).
```

chrisbra commented on runtime/doc/options.txt:
```diff
@@ -2085,6 +2085,28 @@ A jump table for the options with a short description can be found at |Q_op|.
+		name of a function or a |Funcref|.  For |Funcref| values,
+		spaces must be escaped with a backslash ('\'), and commas with
+		double backslashes ('\\').
+		If the {func} result includes 'refresh' set to 'always',
```
```suggestion
		If the Dict returned by the {func} includes {'refresh': 'always'},
```

chrisbra commented on src/optionstr.c:
```diff
@@ -1551,57 +1551,67 @@ did_set_commentstring(optset_T *args)
 }
 #endif
 
+#define MAX_TOKENS 100  // Maximum unique substrings to check
```
this seems unused

girishji replied:
Committed change. Thanks!

girishji committed:
Update runtime/doc/options.txt

Co-authored-by: Christian Brabandt <cb@256bit.org>

girishji committed:
Update runtime/doc/options.txt

Co-authored-by: Christian Brabandt <cb@256bit.org>

girishji committed:
Remove unused var

M  src/optionstr.c

chrisbra commented on src/insexpand.c:
```diff
@@ -4412,6 +4464,51 @@ get_next_default_completion(ins_compl_next_state_T *st, pos_T *start_pos)
+	if (name)
+	    vim_free(name);
+	name = vim_strnsave(buf, slen);
+	if (name)
```
```suggestion
	if (name && option_set_callback_func(name, &cb)
	    return &cb;
```

girishji replied:
refactored: https://github.com/vim/vim/pull/17065/commits/1a574bc08eb4b427a601b491d67bc23ea1763a89

chrisbra commented on src/insexpand.c:
```diff
@@ -4412,6 +4464,51 @@ get_next_default_completion(ins_compl_next_state_T *st, pos_T *start_pos)
+    static callback_T	cb;
+    char_u		buf[LSIZE];
+    int			slen;
+    static char_u	*name;
```
```suggestion
    static char_u	*name = NULL;
```

And then below always use `vim_free(name)` ?

girishji replied:
> And then below always use vim_free(name) ?

There is some bug in option_set_callback_func(). Before, I had `name` as non-static, and it was getting freed inside `free_callback()` because `cb_free_name` was set. But it resulted in memory leak according one of your CI checks. So, I changed to this strategy to pass the check. I will try to make it non-static again and test it out.

girishji committed:
Merge branch 'vim:master' into pr_func_compl

girishji committed:
Update src/insexpand.c

Co-authored-by: Christian Brabandt <cb@256bit.org>

girishji commented on src/optionstr.c:
```diff
@@ -1552,56 +1552,64 @@ did_set_commentstring(optset_T *args)
+	while (*p && (*p != ',' || escape) && buf_ptr < (buffer + LSIZE - 1))
 	{
-	    if (s[-1] == 'k' || s[-1] == 's')
+	    if (*p == '\\' && *(p + 1) == ',')
```
This is not necessary. In optionsstr.c, backslashes before spaces are already stripped. A double backslash (\\) is interpreted as a single backslash, which is required before a comma. The `\\` shown above represents an escaped single backslash. Refer to the tests for confirmation. For example, the following test passes:
 ```
  set complete+=ffunction('foo'\\,\ [10\\,\ 20])
```

girishji committed:
Refactor

Removed static variable, and code associated with it.
Easier to read now. However, function name is not visible
inside callback during debugging.

M  src/insexpand.c

zeertzjq commented on runtime/doc/options.txt:
```diff
@@ -2085,6 +2085,28 @@ A jump table for the options with a short description can be found at |Q_op|.
+		name of a function or a |Funcref|.  For |Funcref| values,
+		spaces must be escaped with a backslash ('\'), and commas with
+		double backslashes ('\\') (see |option-backslash|).
+		If the Dict returned by the {func} includes {'refresh': 'always'},
```
```suggestion
		If the Dict returned by the {func} includes {"refresh": "always"},
```

girishji committed:
Update runtime/doc/options.txt

Co-authored-by: zeertzjq <zeertzjq@outlook.com>

glepnir commented on src/insexpand.c:
```diff
@@ -208,6 +212,10 @@ static int	  compl_selected_item = -1;
 
 static int	  *compl_fuzzy_scores;
 
+static int	  *cpt_func_refresh_always; // array indicating which 'cpt' functions have 'refresh:always' set
```
```suggestion
static int	  *cpt_func_refresh_always;  // array indicating which 'cpt' functions have 'refresh:always' set
```

glepnir commented on src/insexpand.c:
```diff
@@ -1352,6 +1368,7 @@ ins_compl_build_pum(void)
     int		compl_no_select = (cur_cot_flags & COT_NOSELECT) != 0;
     int		fuzzy_filter = (cur_cot_flags & COT_FUZZY) != 0;
     int		fuzzy_sort = fuzzy_filter && !(cur_cot_flags & COT_NOSORT);
+
```
```suggestion
```

glepnir commented on src/insexpand.c:
```diff
@@ -6130,3 +6247,221 @@ spell_back_to_badword(void)
+cpt_compl_src_clear(void)
+{
+    vim_free(cpt_func_refresh_always);
+    cpt_func_refresh_always = NULL;
```
```suggestion
    VIM_CLEAR(cpt_func_refresh_always);
```

glepnir commented on src/optionstr.c:
```diff
@@ -1552,56 +1552,64 @@ did_set_commentstring(optset_T *args)
+	}
+
+	// Skip comma and spaces
+	while (*p == ',' || *p == ' ') p++;
```
```suggestion
	while (*p == ',' || *p == ' ')
	    p++;
```

glepnir commented on src/optionstr.c:
```diff
@@ -1552,56 +1552,64 @@ did_set_commentstring(optset_T *args)
 {
     char_u	**varp = (char_u **)args->os_varp;
-    char_u	*s;
+    char_u	*p;
```
```suggestion
    char_u	*p = NULL;
```
always declare a variable with a default value. also in some other functions.

girishji committed:
Add/modify test

To increase coverage

M  src/testdir/test_ins_complete.vim
M  src/testdir/test_options.vim

girishji committed:
Update src/insexpand.c

Co-authored-by: glepnir <glephunter@gmail.com>

girishji committed:
Update src/insexpand.c

Co-authored-by: glepnir <glephunter@gmail.com>

girishji committed:
Update src/insexpand.c

Co-authored-by: glepnir <glephunter@gmail.com>

girishji committed:
Update src/optionstr.c

Co-authored-by: glepnir <glephunter@gmail.com>

girishji committed:
Update src/optionstr.c

Co-authored-by: glepnir <glephunter@gmail.com>
